packages/headless/src/controllers/core/search-box/headless-core-search-box.ts

```diff
@@ -0,0 +1,271 @@
+   */
+  isLoading: boolean;
+
+  /**
+   * Determines if a query suggest request is in progress.
+   */
+  isLoadingSuggestions: boolean;
+}
+
+export interface Suggestion {
+  /**
+   * The suggestion after applying any `highlightOptions`.
+   */
+  highlightedValue: string;
+
+  /**
+   * The suggestion text.
+   */
+  rawValue: string;
+}
+
+/**
+ * The `SearchBox` headless controller offers a high-level interface for designing a common search box UI controller
+ * with [highlighting for query suggestions](https://docs.coveo.com/en/headless/latest/usage/highlighting/).
+ */
+export interface CoreSearchBox extends Controller {
+  /**
+   * Updates the search box text value and shows the suggestions for that value.
+   *
+   * @param value - The string value to update the search box with.
+   */
+  updateText(value: string): void;
+
+  /**
+   * Clears the search box text and the suggestions.
+   */
+  clear(): void;
+
+  /**
+   * Shows the suggestions for the current search box value.
+   */
+  showSuggestions(): void;
+
+  /**
+   * Selects a suggestion and calls `submit`.
+   *
+   * @param value - The string value of the suggestion to select
+   */
+  selectSuggestion(value: string): void;
+
+  /**
+   * Deselects all facets and triggers a search query.
+   */
+  submit(): void;
+
+  /**
+   * The state of the `SearchBox` controller.
+   */
+  state: SearchBoxState;
+}
+
+export function buildCoreSearchBox(
+  engine: CoreEngine,
+  props: CoreSearchBoxProps = {}
+): CoreSearchBox {
+  if (!loadSearchBoxReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+  const getState = () => engine.state;
+
+  const id = props.options?.id || randomID('search_box');
+  const options: Required<SearchBoxOptions> = {
+    id,
+    highlightOptions: {...props.options?.highlightOptions},
+    ...defaultSearchBoxOptions,
+    ...props.options,
+  };
+
+  validateOptions(engine, searchBoxOptionsSchema, options, 'buildSearchBox');
+  dispatch(registerQuerySetQuery({id, query: engine.state.query.q}));
+  if (options.numberOfSuggestions) {
+    dispatch(
+      registerQuerySuggest({
+        id,
+        q: engine.state.query.q,
+        count: options.numberOfSuggestions,
+      })
+    );
+  }
+
+  const getValue = () => engine.state.querySet[options.id];
+
+  const performSearch = async (analytics: SearchAction) => {
+    const {enableQuerySyntax} = options;
+
+    dispatch(prepareForSearchWithQuery({q: getValue(), enableQuerySyntax}));
+
+    if (props.executeSearchActionCreator) {
+      await dispatch(props.executeSearchActionCreator(analytics));
+    }
+  };
+
+  return {
+    ...controller,
+
+    updateText(value: string) {
+      dispatch(updateQuerySetQuery({id, query: value}));
+      this.showSuggestions();
+    },
+
+    clear() {
+      dispatch(updateQuerySetQuery({id, query: ''}));
+      dispatch(clearQuerySuggest({id}));
+    },
+
+    showSuggestions() {
```

Comment:
Similar to my above comment, we should either make `props.fetchQuerySuggestionsActionCreator` required, and remove the check on line `202`, or move the `showSuggestions()` function to each specific implementation.

If the insight panel backend service does not yet support query suggestions, we could add a no-op action with a `TODO` to implement it.